presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonToRowCast.java

```diff
@@ -0,0 +1,176 @@
+
+public class JsonToRowCast
+        extends SqlOperator
+{
+    public static final JsonToRowCast JSON_TO_ROW = new JsonToRowCast();
+    private static final MethodHandle METHOD_HANDLE = methodHandle(JsonToRowCast.class, "toRow", RowType.class, BlockBuilderAppender[].class, Optional.class, Optional.class, ConnectorSession.class, Slice.class);
+
+    private JsonToRowCast()
+    {
+        super(OperatorType.CAST,
+                ImmutableList.of(withVariadicBound("T", "row")),
+                ImmutableList.of(),
+                parseTypeSignature("T"),
+                ImmutableList.of(parseTypeSignature(StandardTypes.JSON)));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(arity == 1, "Expected arity to be 1");
+        Type type = boundVariables.getTypeVariable("T");
+        RowType rowType = (RowType) type;
+        checkCondition(canCastFromJson(rowType), INVALID_CAST_ARGUMENT, "Cannot cast JSON to %s", rowType);
+
+        List<RowField> rowFields = rowType.getFields();
+        BlockBuilderAppender[] fieldAppenders = new BlockBuilderAppender[rowFields.size()];
+        for (int i = 0; i < fieldAppenders.length; i++) {
+            fieldAppenders[i] = createBlockBuilderAppender(rowFields.get(i).getType());
+        }
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(rowType).bindTo(fieldAppenders).bindTo(getFieldNameToIndex(rowFields)).bindTo(getFieldNames(rowFields));
+        return new ScalarFunctionImplementation(true, ImmutableList.of(false), methodHandle, isDeterministic());
+    }
+
+    @UsedByGeneratedCode
+    public static Block toRow(RowType rowType, BlockBuilderAppender[] fieldAppenders, Optional<Map<String, Integer>> fieldNameToIndex, Optional<List<String>> fieldNames, ConnectorSession connectorSession, Slice json)
+    {
+        try (JsonParser jsonParser = createJsonParser(JSON_FACTORY, json)) {
+            jsonParser.nextToken();
+            if (jsonParser.getCurrentToken() == JsonToken.VALUE_NULL) {
+                return null;
+            }
+
+            if (jsonParser.getCurrentToken() != START_ARRAY && jsonParser.getCurrentToken() != START_OBJECT) {
+                throw new JsonCastException(format("Expected a json array or object, but got %s", jsonParser.getText()));
+            }
+
+            // NOTE: We can use RowBlockBuilderAppender as it has the same logic with the following code.
+            // We didn't do so because it requested to create a RowBlockBuilder even for NULL value and
+            // the JSON value is not valid, which is not cheap.
+            // TODO: We should use RowBlockBuilderAppender once CAST function support to take a BlockBuilder
+            // as input or it supports cachedInstanceFactory.
+            BlockBuilder rowBlockBuilder = rowType.createBlockBuilder(new BlockBuilderStatus(), 1);
```

Comment:
Given the complicated logic of JSON->ROW cast, I think it makes sense to just use the `rowBlockBuilderAppender` at the cost of always creating a new `RowBlock`.